bgpd/bgp_community.c

```diff
@@ -910,3 +910,142 @@ void community_finish(void)
+		if (memcmp(com1->val, com2->val, com1->size * 4) == 0)
+			return true;
+	return false;
+}
+
+static struct community *bgp_aggr_community_lookup(
+						struct bgp_aggregate *aggregate,
+						struct community *community)
+{
+	return hash_lookup(aggregate->community_hash, community);
+}
+
+static void *bgp_aggr_communty_hash_alloc(void *p)
+{
+	struct community *ref = (struct community *)p;
+	struct community *community = NULL;
+
+	community = community_dup(ref);
+	return community;
+}
+
+static void bgp_aggr_community_prepare(struct hash_backet *hb, void *arg)
+{
+	struct community *commerge = NULL;
+	struct community *hb_community = hb->data;
+	struct community **aggr_community = arg;
+
+	if (*aggr_community) {
+		commerge = community_merge(*aggr_community, hb_community);
+		*aggr_community = community_uniq_sort(commerge);
+		community_free(&commerge);
+	} else
+		*aggr_community = community_dup(hb_community);
+}
+
+void bgp_aggr_community_remove(void *arg)
+{
+	struct community *community = arg;
+
+	community_free(&community);
+}
+
+void bgp_compute_aggregate_community(struct bgp_aggregate *aggregate,
+				     struct community *community)
+{
+	struct community *aggr_community = NULL;
+
+	if ((aggregate == NULL) || (community == NULL))
+		return;
+
+	/* Create hash if not already created.
+	 */
+	if (aggregate->community_hash == NULL) {
+		aggregate->community_hash = hash_create(
+					bgp_aggr_community_hash_key,
+					bgp_aggr_community_hash_cmp,
+					"BGP Aggregator community hash");
+		assert(aggregate->community_hash != NULL);
+	}
+
+	aggr_community = bgp_aggr_community_lookup(aggregate, community);
+	if (aggr_community == NULL) {
+		/* Insert community into hash.
+		 */
+		aggr_community = hash_get(aggregate->community_hash, community,
+					  bgp_aggr_communty_hash_alloc);
+		assert(aggr_community != NULL);
+
+		/* Re-compute aggregate's community.
+		 */
+		if (aggregate->community) {
+			community_free(&aggregate->community);
+			aggregate->community = NULL;
+		}
+
+		hash_iterate(aggregate->community_hash,
+			     bgp_aggr_community_prepare,
+			     &aggregate->community);
+	}
+
+	/* Increment refernce counter.
+	 */
+	aggr_community->refcnt++;
+}
+
+void bgp_remove_community_from_aggregate(struct bgp_aggregate *aggregate,
+					 struct community *community)
+{
+	struct community *aggr_community = NULL;
+	struct community *ret_comm = NULL;
+
+	if ((aggregate == NULL) || (community == NULL))
+		return;
+
+	if (aggregate->community_hash == NULL)
+		return;
+
+	/* Look-up the community in the hash.
+	 */
+	aggr_community = bgp_aggr_community_lookup(aggregate, community);
+	if (aggr_community) {
+		aggr_community->refcnt--;
+
+		if (aggr_community->refcnt == 0) {
+			ret_comm = hash_release(aggregate->community_hash,
+						aggr_community);
+			assert(ret_comm != NULL);
+			community_free(&ret_comm);
+
+			community_free(&aggregate->community);
+			aggregate->community = NULL;
+
+			if (aggregate->community_hash->count > 0)
```

Comment:
Just call hash_iterate it will do the right thing. No need to check for a count.

Reply:
I've removed this check.